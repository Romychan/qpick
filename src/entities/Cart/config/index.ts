/** The name of the storage for storing data in the cart */
export const CART_STORAGE_NAME = 'cart';
