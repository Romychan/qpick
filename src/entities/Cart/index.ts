export { MOCK_CART, MOCK_CART_PRODUCT } from './lib/mock';

export { CartSummary } from './ui/CartSummary/CartSummary';

export { CartContext, CartContextProvider } from './model/context';
export { useCartStore } from './model/hooks';
