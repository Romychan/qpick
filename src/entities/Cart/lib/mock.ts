// eslint-disable-next-line @conarti/feature-sliced/layers-slices
import { Product } from '~/entities/Product/@x';

import { CartItem } from '../model/types';

/** Mock product for the cart */
export const MOCK_CART_PRODUCT: Product = {
  id: 1,
  title: 'Product',
  rate: 4.2,
  image: '/images/catalog/headphone-1.png',
  price: 1200,
  oldPrice: 1300,
  description: 'Test',
  category: 'Наушники',
};

/** List of mock cart items */
export const MOCK_CART: CartItem[] = [
  {
    product: MOCK_CART_PRODUCT,
    quantity: 2,
  },
];
