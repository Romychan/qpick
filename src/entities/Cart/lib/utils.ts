// eslint-disable-next-line @conarti/feature-sliced/layers-slices
import { Product } from '~/entities/Product/@x';

import { CartItem } from '../model/types';

/**
 * Add a product to the cart
 *
 * @param cartItems The current state of the cart
 * @param product Product to add to cart
 * @returns Updated state of the product cart
 */
export const addToCart = (cartItems: CartItem[], product: Product) => {
  const productInCart = cartItems.find(
    (item) => item.product.id === product.id,
  );

  if (!productInCart) {
    return [...cartItems, { product, quantity: 1 }];
  } else {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity + 1 }
        : item,
    );
  }
};

/**
 * Delete a product from the cart
 *
 * @param cartItems The current state of the cart
 * @param id The ID of the product to be deleted from the cart
 * @returns Updated state of the product cart
 */
export const deleteFromCart = (cartItems: CartItem[], id: number) => {
  return cartItems.filter((item) => item.product.id !== id);
};

/**
 * Increase the number of product in the cart
 *
 * @param cartItems The current state of the cart
 * @param id Product ID to increase the quantity
 * @returns Updated state of the product cart
 */
export const increaseProductQuantity = (cartItems: CartItem[], id: number) => {
  const productInCart = cartItems.find((item) => item.product.id === id);

  if (productInCart) {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity + 1 }
        : item,
    );
  }

  return cartItems;
};

/**
 * Decrease the number of product in the cart
 *
 * @param cartItems The current state of the cart
 * @param id Product ID to decrease the quantity
 * @returns Updated state of the product cart
 */
export const decreaseProductQuantity = (cartItems: CartItem[], id: number) => {
  const productInCart = cartItems.find((item) => item.product.id === id);

  if (productInCart) {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity - 1 }
        : item,
    );
  }

  return cartItems;
};
