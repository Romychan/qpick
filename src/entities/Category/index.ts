export type { Category } from './model/types';
export { MAIN_CATEGORIES } from './model/constants';
