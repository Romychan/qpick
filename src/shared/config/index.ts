export const MAIN_ROUTE = '/';

export const CART_ROUTE = '/cart';

export const FAVORITE_ROUTE = '/favorite';

export const CONTACT_ROUTE = '/contact';

export const TERMS_ROUTE = '/terms';

export const OVERVIEW_ROUTE = '/overview';

export const ORDER_ROUTE = '/order';
