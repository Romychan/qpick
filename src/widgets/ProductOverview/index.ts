export { ProductOverview } from './ui/ProductOverview';
