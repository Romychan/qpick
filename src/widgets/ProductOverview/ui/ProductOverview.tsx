import { Navigate, useNavigate, useParams } from 'react-router-dom';

import { AddToCart } from '~/features/AddToCart';

import { MOCK_PRODUCTS, ProductDetails } from '~/entities/Product';

import { Drawer } from '~/shared/ui/Drawer';

import styles from './ProductOverview.module.scss';

/** This component is used to display detailed product information in a modal window */
export const ProductOverview = () => {
  const { productId } = useParams();
  const navigate = useNavigate();

  const currentProduct = MOCK_PRODUCTS.find(
    (element) => element.id === Number(productId),
  );

  if (!productId || !currentProduct) {
    return <Navigate to="/" />;
  }

  return (
    <Drawer
      description={`Каталог / ${currentProduct.category} / ${currentProduct.title}`}
      className={styles.container}
      onClose={() => navigate('/')}
      isVisible
    >
      <ProductDetails
        product={currentProduct}
        actions={<AddToCart product={currentProduct} />}
      />
    </Drawer>
  );
};
