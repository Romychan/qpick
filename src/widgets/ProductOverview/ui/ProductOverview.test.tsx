import { screen } from '@testing-library/react';
import { Route, Routes } from 'react-router-dom';

import { CartContextProvider } from '~/entities/Cart';

import { renderWithRouter } from '~/shared/lib/tests';
import { OVERVIEW_ROUTE } from '~/shared/config';

import { ProductOverview } from './ProductOverview';

describe('ProductOverview', () => {
  it('should render correctly', () => {
    const productId = 1;

    renderWithRouter(
      <Routes>
        <Route
          path={`${OVERVIEW_ROUTE}/:productId`}
          element={
            <CartContextProvider>
              <ProductOverview />
            </CartContextProvider>
          }
        />
      </Routes>,
      `${OVERVIEW_ROUTE}/${productId}`,
    );

    const drawerElement = screen.getByTestId('drawer');
    const detailsElement = screen.getByTestId('product-details');

    expect(drawerElement).toBeInTheDocument();
    expect(detailsElement).toBeInTheDocument();
  });

  it('should not show the product details if an invalid ID is specified', () => {
    const productId = 'TestTestTest123';

    renderWithRouter(
      <Routes>
        <Route
          path={`${OVERVIEW_ROUTE}/:productId`}
          element={
            <CartContextProvider>
              <ProductOverview />
            </CartContextProvider>
          }
        />
        <Route path="/" element="Empty" />
      </Routes>,
      `${OVERVIEW_ROUTE}/${productId}`,
    );

    const drawerElement = screen.queryByTestId('drawer');
    const detailsElement = screen.queryByTestId('product-details');

    expect(drawerElement).not.toBeInTheDocument();
    expect(detailsElement).not.toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const productId = 1;

    const container = renderWithRouter(
      <Routes>
        <Route
          path={`${OVERVIEW_ROUTE}/:productId`}
          element={
            <CartContextProvider>
              <ProductOverview />
            </CartContextProvider>
          }
        />
      </Routes>,
      `${OVERVIEW_ROUTE}/${productId}`,
    );

    expect(container).toMatchSnapshot();
  });
});
