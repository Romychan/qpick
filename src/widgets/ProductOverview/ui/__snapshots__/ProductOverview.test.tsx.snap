// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductOverview > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_backdrop_626b11 _center_626b11 _filled_626b11"
        data-testid="drawer-backdrop"
      />
      <div
        class="_container_626b11 _center_626b11 _container_415d6e"
        data-testid="drawer"
      >
        <div
          class="_header_626b11"
        >
          <div
            class="_information_626b11"
          >
            <h3
              class="_title_626b11"
            />
            <p
              class="_description_626b11"
            >
              Каталог / Наушники / Apple BYZ S852I
            </p>
          </div>
          <button
            aria-label="Закрыть модальное окно"
            class="_button_0302f3 _tertiary_0302f3 _sm_0302f3 _icon_0302f3"
            data-testid="drawer-close"
            title="Закрыть"
            type="button"
          >
            <svg
              class="icon icon-close"
              data-testid="icon"
              height="16"
              width="16"
            >
              <use
                xlink:href="/images/icons.svg#close"
              />
            </svg>
          </button>
        </div>
        <div
          class="_body_626b11"
        >
          <div
            class="_content_626b11"
          >
            <div
              class="_container_ff5a21"
              data-testid="product-details"
            >
              <div
                class="_image_ff5a21"
              >
                <img
                  alt="Apple BYZ S852I"
                  height="270"
                  src="/images/catalog/headphone-1.png"
                  width="250"
                />
              </div>
              <div
                class="_content_ff5a21"
              >
                <div
                  class="_information_ff5a21"
                >
                  <h1
                    class="_name_ff5a21"
                  >
                    Apple BYZ S852I
                  </h1>
                  <div
                    class="_container_861903"
                    data-testid="rating"
                  >
                    <svg
                      class="icon icon-star-filled _icon_861903 filled"
                      data-testid="icon"
                      height="24"
                      width="24"
                    >
                      <use
                        xlink:href="/images/icons.svg#star-filled"
                      />
                    </svg>
                    <p
                      class="_text_861903"
                    >
                      4.7
                    </p>
                  </div>
                </div>
                <div
                  class="_information_ff5a21"
                >
                  <div
                    class="_container_5a3e99 _horizontal_5a3e99 _lg_5a3e99"
                    data-testid="price"
                  >
                    <span
                      class="_price_5a3e99"
                    >
                      2 927 ₽
                       
                    </span>
                    <span
                      class="_old_5a3e99"
                    >
                      3 527 ₽
                       
                    </span>
                  </div>
                  <div
                    class="_actions_ff5a21"
                  >
                    <button
                      class="_button_0302f3 _button_9e4ee0 _link-primary_0302f3 _md_0302f3"
                      data-testid="button-add-to-cart"
                      type="button"
                    >
                      Купить
                    </button>
                  </div>
                </div>
                <div
                  class="_information_ff5a21"
                >
                  <div
                    class="_specification_ff5a21"
                  >
                    <p
                      class="_subtitle_ff5a21"
                    >
                      Описание
                    </p>
                    <p
                      class="_text_ff5a21"
                    >
                      Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_backdrop_626b11 _center_626b11 _filled_626b11"
      data-testid="drawer-backdrop"
    />
    <div
      class="_container_626b11 _center_626b11 _container_415d6e"
      data-testid="drawer"
    >
      <div
        class="_header_626b11"
      >
        <div
          class="_information_626b11"
        >
          <h3
            class="_title_626b11"
          />
          <p
            class="_description_626b11"
          >
            Каталог / Наушники / Apple BYZ S852I
          </p>
        </div>
        <button
          aria-label="Закрыть модальное окно"
          class="_button_0302f3 _tertiary_0302f3 _sm_0302f3 _icon_0302f3"
          data-testid="drawer-close"
          title="Закрыть"
          type="button"
        >
          <svg
            class="icon icon-close"
            data-testid="icon"
            height="16"
            width="16"
          >
            <use
              xlink:href="/images/icons.svg#close"
            />
          </svg>
        </button>
      </div>
      <div
        class="_body_626b11"
      >
        <div
          class="_content_626b11"
        >
          <div
            class="_container_ff5a21"
            data-testid="product-details"
          >
            <div
              class="_image_ff5a21"
            >
              <img
                alt="Apple BYZ S852I"
                height="270"
                src="/images/catalog/headphone-1.png"
                width="250"
              />
            </div>
            <div
              class="_content_ff5a21"
            >
              <div
                class="_information_ff5a21"
              >
                <h1
                  class="_name_ff5a21"
                >
                  Apple BYZ S852I
                </h1>
                <div
                  class="_container_861903"
                  data-testid="rating"
                >
                  <svg
                    class="icon icon-star-filled _icon_861903 filled"
                    data-testid="icon"
                    height="24"
                    width="24"
                  >
                    <use
                      xlink:href="/images/icons.svg#star-filled"
                    />
                  </svg>
                  <p
                    class="_text_861903"
                  >
                    4.7
                  </p>
                </div>
              </div>
              <div
                class="_information_ff5a21"
              >
                <div
                  class="_container_5a3e99 _horizontal_5a3e99 _lg_5a3e99"
                  data-testid="price"
                >
                  <span
                    class="_price_5a3e99"
                  >
                    2 927 ₽
                     
                  </span>
                  <span
                    class="_old_5a3e99"
                  >
                    3 527 ₽
                     
                  </span>
                </div>
                <div
                  class="_actions_ff5a21"
                >
                  <button
                    class="_button_0302f3 _button_9e4ee0 _link-primary_0302f3 _md_0302f3"
                    data-testid="button-add-to-cart"
                    type="button"
                  >
                    Купить
                  </button>
                </div>
              </div>
              <div
                class="_information_ff5a21"
              >
                <div
                  class="_specification_ff5a21"
                >
                  <p
                    class="_subtitle_ff5a21"
                  >
                    Описание
                  </p>
                  <p
                    class="_text_ff5a21"
                  >
                    Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
